File: javascript/fizzbuzz/fizzbuzz.js
const FrameEvent = require("./framelang/framelang");

// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

class FizzBuzz {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sBegin_;
        this.#compartment = new FizzBuzzCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    start() {
        let e = FrameEvent(">>",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sBegin_:
                this.#sBegin_(e);
                break;
            case this.#sFizz_:
                this.#sFizz_(e);
                break;
            case this.#sBuzz_:
                this.#sBuzz_(e);
                break;
            case this.#sDigit_:
                this.#sDigit_(e);
                break;
            case this.#sEnd_:
                this.#sEnd_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sBegin_(e) {
        switch (e._message) {
            case ">>":
                {
                // start
                let compartment =  new FizzBuzzCompartment(this.#sFizz_);
                
                compartment.EnterArgs["i"] = 1;
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sFizz_(e) {
        switch (e._message) {
            case ">":
                {
                if (this.gt_100_do((e._parameters["i"]))) {
                    // i > 100
                    let compartment =  new FizzBuzzCompartment(this.#sEnd_);
                    
                    
                    this.#transition(compartment);
                    return;
                }
                if (this.mod3_eq0_do((e._parameters["i"]))) {
                    this.print_do("Fizz");
                    // i % 3 == 0
                    let compartment =  new FizzBuzzCompartment(this.#sBuzz_);
                    
                    compartment.EnterArgs["i"] = e._parameters["i"];
                    compartment.EnterArgs["fizzed"] = true;
                    
                    this.#transition(compartment);
                } else {
                    // i % 3 != 0
                    let compartment =  new FizzBuzzCompartment(this.#sBuzz_);
                    
                    compartment.EnterArgs["i"] = e._parameters["i"];
                    compartment.EnterArgs["fizzed"] = false;
                    
                    this.#transition(compartment);
                }
                return;
                }
                
        }
    }
    
    #sBuzz_(e) {
        switch (e._message) {
            case ">":
                {
                if (this.mod5_eq0_do((e._parameters["i"]))) {
                    this.print_do("Buzz");
                    // i % 5 == 0
                    this.#compartment.ExitArgs["output"] = " ";
                    let compartment =  new FizzBuzzCompartment(this.#sFizz_);
                    
                    compartment.EnterArgs["i"] = this.plus_1_do(e._parameters["i"]);
                    
                    this.#transition(compartment);
                    return;
                }
                if ((e._parameters["fizzed"])) {
                    // fizzed
                    this.#compartment.ExitArgs["output"] = " ";
                    let compartment =  new FizzBuzzCompartment(this.#sFizz_);
                    
                    compartment.EnterArgs["i"] = this.plus_1_do(e._parameters["i"]);
                    
                    this.#transition(compartment);
                    return;
                }
                // ! mod3 | mod5
                this.#compartment.ExitArgs["output"] = "";
                let compartment =  new FizzBuzzCompartment(this.#sDigit_);
                
                compartment.EnterArgs["i"] = e._parameters["i"];
                
                this.#transition(compartment);
                return;
                }
                
            case "<":
                {
                this.print_do((e._parameters["output"]));
                return;
                }
                
        }
    }
    
    #sDigit_(e) {
        switch (e._message) {
            case ">":
                {
                this.print_do((e._parameters["i"]).toString());
                this.print_do(" ");
                // loop
                let compartment =  new FizzBuzzCompartment(this.#sFizz_);
                
                compartment.EnterArgs["i"] = this.plus_1_do(e._parameters["i"]);
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sEnd_(e) {
        switch (e._message) {
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    print_do(msg) { throw new Error('Action not implemented.'); }
    gt_100_do(i) { throw new Error('Action not implemented.'); }
    mod3_eq0_do(i) { throw new Error('Action not implemented.'); }
    mod5_eq0_do(i) { throw new Error('Action not implemented.'); }
    plus_1_do(i) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};


//=============== Compartment ==============//

class FizzBuzzCompartment {
  constructor(state) {
    this.state = state;
  }

  StateArgs = {};
  StateVars = {};
  EnterArgs = {};
  ExitArgs = {};
  _forwardEvent = FrameEvent.call(this);
}

class FizzBuzzController extends FizzBuzz {
  constructor() {
    super();
    
  }
  print_do(msg) {
    process.stdout.write(msg.toString());
  }
  gt_100_do(i) {
    return i > 100;
  }
  mod3_eq0_do(i) {
    return i % 3 == 0;
  }
  mod5_eq0_do(i) {
    return i % 5 == 0;
  }
  plus_1_do(i) {
    return i + 1;
  }
}

module.exports = FizzBuzzController

File: java/fizzbuzz/fizzbuzz.java
import java.util.*;

// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

enum FizzBuzzState {
	BEGIN(0), FIZZ(1), BUZZ(2), DIGIT(3), END(4);

	public final int value;

	private FizzBuzzState(int value) {
		this.value = value;
	}

	public int getValue() {
		return value;
	}
}

class FizzBuzz {
	private FizzBuzzCompartment _compartment_;
	private FizzBuzzCompartment _nextCompartment_;

	FizzBuzz() {
		this._compartment_ = new FizzBuzzCompartment(this._state_);
		this._nextCompartment_ = null;
		
		FrameEvent e = new FrameEvent(">", null);
		this._mux_(e);
	}

	// ===================== Interface Block ===================//

	public void start() {
		FrameEvent e = new FrameEvent(">>", null);
		this._mux_(e);
	}

	// ====================== Multiplexer ====================//

	private void _mux_(FrameEvent e) {

		if (this._compartment_.state == FizzBuzzState.BEGIN.getValue()) {
			this._sBegin_(e);
		} else if(this._compartment_.state == FizzBuzzState.FIZZ.getValue()) {
			this._sFizz_(e);
		} else if(this._compartment_.state == FizzBuzzState.BUZZ.getValue()) {
			this._sBuzz_(e);
		} else if(this._compartment_.state == FizzBuzzState.DIGIT.getValue()) {
			this._sDigit_(e);
		} else if(this._compartment_.state == FizzBuzzState.END.getValue()) {
			this._sEnd_(e);
		}

		if (this._nextCompartment_ != null) {
			FizzBuzzCompartment nextCompartment = this._nextCompartment_;
			this._nextCompartment_ = null;
			if (nextCompartment._forwardEvent != null && nextCompartment._forwardEvent._message == ">") {
				this._mux_(new FrameEvent("<", this._compartment_.exitArgs));
				this._compartment_ = nextCompartment;
				this._mux_(nextCompartment._forwardEvent);
			} else {
				this._doTransition_(nextCompartment);
				if (nextCompartment._forwardEvent != null) {
					this._mux_(nextCompartment._forwardEvent);
				}
			}
			nextCompartment._forwardEvent = null;
		}
	}

	private void _doTransition_(FizzBuzzCompartment nextCompartment) {
		this._mux_(new FrameEvent("<", this._compartment_.exitArgs));
		this._compartment_ = nextCompartment;
		this._mux_(new FrameEvent(">", this._compartment_.enterArgs));
	}

	// ===================== Machine Block ===================// // see working demo
	// at codepen.io:
	// https://codepen.io/frame-lang/pen/e96568889939d3bc112b16528900c352
	// try out on http://framepiler.frame-lang.org

	private void _sBegin_(FrameEvent e) {
		if (e._message == ">>") {
			FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.FIZZ.getValue());
			compartment.enterArgs.put("i", 1);
			this._transition_(compartment);
			return;
		}
	}

	private void _sFizz_(FrameEvent e) {
		if (e._message == ">") {
			if (gt_100_do(((int) e._parameters.get("i")))) {
				// i > 100
				FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.END.getValue());
				this._transition_(compartment);
				return;
			}
			if (mod3_eq0_do(((int) e._parameters.get("i")))) {
				print_do("Fizz");
				// i % 3 == 0
				FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.BUZZ.getValue());
				compartment.enterArgs.put("i", e._parameters.get("i"));
                compartment.enterArgs.put("fizzed", true);
				this._transition_(compartment);
				return;
			} else {
				// i % 3 != 0
				FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.BUZZ.getValue());
				compartment.enterArgs.put("i", e._parameters.get("i"));
                compartment.enterArgs.put("fizzed", false);
				this._transition_(compartment);
			}
			return;
		}
	}

	private void _sBuzz_(FrameEvent e) {
		if (e._message == ">") {
			if (mod5_eq0_do(((int) e._parameters.get("i")))) {
				print_do("Buzz");
				// i % 5 == 0
				this._compartment_.exitArgs.put("output", " ");
				FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.FIZZ.getValue());
				compartment.enterArgs.put("i", this.plus_1_do((int) e._parameters.get("i")));
				this._transition_(compartment);
				return;
			}
			if (((boolean) e._parameters.get("fizzed"))) {
				// fizzed
				this._compartment_.exitArgs.put("output", " ");
				FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.FIZZ.getValue());
				compartment.enterArgs.put("i", this.plus_1_do((int) e._parameters.get("i")));
				this._transition_(compartment);
				return;
			}
			// ! mod3 | mod5
			this._compartment_.exitArgs.put("output", " ");
			FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.DIGIT.getValue());
			compartment.enterArgs.put("i", this.plus_1_do((int) e._parameters.get("i")));
			this._transition_(compartment);
			return;
		} else if (e._message == "<") {
			print_do(((String) e._parameters.get("output")));
			return;
		}
	}

	private void _sDigit_(FrameEvent e) {
		if (e._message == ">") {
			print_do((String.valueOf((int) e._parameters.get("i"))));
			print_do(" ");
			// loop
			FizzBuzzCompartment compartment = new FizzBuzzCompartment(FizzBuzzState.FIZZ.getValue());
			compartment.enterArgs.put("i", this.plus_1_do((int) e._parameters.get("i")));
			this._transition_(compartment);
			return;
		}
	}

	private void _sEnd_(FrameEvent e) {
	}

	// ===================== Actions Block ===================//

	protected void print_do(String msg) {
		throw new UnsupportedOperationException();
	}

	protected boolean gt_100_do(int i) {
		throw new UnsupportedOperationException();
	}

	protected boolean mod3_eq0_do(int i) {
		throw new UnsupportedOperationException();
	}

	protected boolean mod5_eq0_do(int i) {
		throw new UnsupportedOperationException();
	}

	protected int plus_1_do(int i) {
		throw new UnsupportedOperationException();
	}

	// =============== Machinery and Mechanisms ==============//

	private int _state_;

	private void _transition_(FizzBuzzCompartment compartment) {
		this._nextCompartment_ = compartment;
	}

}

//====================Compartment====================//

class FizzBuzzCompartment {

	int state;

	FizzBuzzCompartment(int state) {
		this.state = state;
	}

	HashMap<String, Object> stateArgs = new HashMap<String, Object>();
	HashMap<String, Object> stateVars = new HashMap<String, Object>();
	HashMap<String, Object> enterArgs = new HashMap<String, Object>();
	HashMap<String, Object> exitArgs = new HashMap<String, Object>();

	FrameEvent _forwardEvent = new FrameEvent();
}

class FizzBuzzController extends FizzBuzz {

	FizzBuzzController() {
		super();
	}

	protected void print_do(String msg) {
		System.out.print(msg + " ");
	}

	protected boolean gt_100_do(int i) {
		return i > 100;
	}

	protected boolean mod3_eq0_do(int i) {
		return i % 3 == 0;
	}

	protected boolean mod5_eq0_do(int i) {
		return i % 5 == 0;
	}

	protected int plus_1_do(int i) {
		return i + 1;
	}
}